lib/signs_ui/config/sign_groups.ex

```diff
@@ -0,0 +1,50 @@
+defmodule SignsUi.Config.SignGroups do
+  @moduledoc """
+  Represents the state of all current sign groups.
+  """
+
+  alias SignsUi.Alerts.Alert
+  alias SignsUi.Config.SignGroup
+
+  @type route_id() :: String.t()
+  @type unix_time() :: String.t()
+  @type t() :: %{unix_time() => SignGroup.t() | %{}}
+  @type by_route() :: %{route_id() => t()}
+
+  @doc """
+  Returns a map of expired group timestamps to empty maps.
+  """
+  @spec expired(t(), DateTime.t(), MapSet.t(Alert.id())) :: t()
+  def expired(sign_groups, current_time, alerts) do
+    for {created_at, group} <- sign_groups,
+        SignGroup.expired?(group, current_time, alerts),
+        into: %{} do
+      {created_at, %{}}
+    end
+  end
+
+  @spec from_json(t()) :: t()
+  def from_json(groups) do
+    Map.new(groups, fn {time, group} -> {time, SignGroup.from_json(group)} end)
+  end
+
+  @spec by_route(t()) :: by_route()
+  def by_route(groups) do
+    groups
```

Comment:
I think this is fine, but just as an option, it can be done in a single pass with a reduce, and `Map.update` which has an optional 4th argument if the key is already present:

```ex
Enum.reduce(sign_groups, %{}, fn {time, group}, acc ->
  Map.update(acc, group.route_id, %{time => group}, &(Map.put(&1, time, group)) )
end)
```

Reply:
Just when I thought I had the most elegant solution!